include/xtensor/xhistogram.hpp

```diff
@@ -0,0 +1,307 @@
+     * @param If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2, class E3>
+    inline auto histogram(
+        const E1 &data, const E2 &bin_edges, const E3 &weights, bool density=false
+    )
+    {
+        // counter type
+        using size_type = detail::common_size_type_t<E1,E2,E3>;
+
+        // basic checks
+        #ifdef XTENSOR_ENABLE_ASSERT
+
+            // - shape and size
+            XTENSOR_ASSERT( data     .dimension() == 1 );
+            XTENSOR_ASSERT( weights  .dimension() == 1 );
+            XTENSOR_ASSERT( bin_edges.dimension() == 1 );
+            XTENSOR_ASSERT( weights  .size() == data.size() );
+            XTENSOR_ASSERT( bin_edges.size() >= 2 );
+            // - bin-edges must be sorted
+            for (size_type i = 1 ; i < bin_edges.size() ; ++i)
+            {
+                XTENSOR_ASSERT( bin_edges[i] >= bin_edges[i - 1] );
+            }
+            // - data must be enclosed in the bins
+            XTENSOR_ASSERT( xt::amin(data)[0] >= bin_edges[0] );
+            XTENSOR_ASSERT( xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1] );
+
+        #endif
+
+        // initialize output
+        xt::xtensor<double,1> count = xt::zeros<double>({bin_edges.size()-1});
+
+        // indices that sort "data"
+        xt::xtensor<size_t,1> isort = xt::argsort(data);
+
+        // index of the current bin
+        size_type ibin = 0;
+
+        // fill the histogram: loop over (sorted) data
+        for (auto &idx : isort)
+        {
+            // - proceed to the relevant bin
+            while ( data[idx] >= bin_edges[ibin+1] and ibin < bin_edges.size()-2 )
+            {
+                ++ibin;
+            }
+            // - update the count
+            count[ibin] += weights[idx];
+        }
+
+        // normalize
+        if ( density )
+        {
+            // - size as doubles
+            double N = static_cast<double>(data.size());
+            // - apply normalization
+            for (size_type i = 0 ; i < bin_edges.size()-1 ; ++i)
+            {
+                count[i] /= ( (bin_edges[i+1]-bin_edges[i]) * N );
+            }
+        }
+
+        return count;
+    }
+
+    /**
+     * @ingroup basic_functions
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param The data.
+     * @param The bin-edges.
+     * @param If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2>
+    inline auto histogram(
+        const E1 &data, const E2 &bin_edges, bool density=false
+    )
+    {
+        using value_type = typename E1::value_type;
+
+        xt::xtensor<value_type,1> weights = xt::ones<value_type>({data.size()});
+
+        return histogram(data, bin_edges, weights, density);
```

Comment:
you can pass directly `xt::ones<value_type>({data.size()})` instead of assigning it to a `xtensor`, this will save memory (`xt::ones` does not allocate any memory).